### ISSUE TITLE
Could not complete schema update on fresh Install of 2.7.2-1 with Postgresql 9.6

### ISSUE THREAD

**Issue by sumpfgottheit** [rundeck/rundeck#2354]
Issue type: Bug report
**My Rundeck detail**

* Rundeck version: 2.7.2-1
* install type: rpm
* OS Name/version: Rhel 6.8
* Database: Postgresql 9.6

In preparation of an update from our current installation (Rundek 2.6.8-1, Deployment via war/tomcat, Postgresql 9.3) I prepared our new rundeck environment and came over a strange error.
Rundeck itself seems to work neverthereless, but I don't know if this migration/field or the absence of it will lead to any problems.

**Expected Behavior**

No errors regarding schema update in /var/log/services.log

**Actual Behavior**

The installation works (all puppetized), database is filled and everything looks good. As soon as I have the first connection to the webinterface, I see the following exception in /var/log/services.log:

```
2017-03-06 14:51:39.431:INFO:/:main: Initializing Spring root WebApplicationContext
2017-03-06 14:51:52,722 [main] ERROR spi.SqlExceptionHelper  - ERROR: column am.amcanorder does not exist
  Position: 407
2017-03-06 14:51:52,724 [main] ERROR hbm2ddl.SchemaUpdate  - HHH000299: Could not complete schema update
java.lang.NullPointerException
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:126)
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:112)
        ......
INFO  BootStrap: Starting Rundeck 2.7.2-1 (2017-02-03) ...
INFO  BootStrap: using rdeck.base config property: /var/lib/rundeck
INFO  BootStrap: loaded configuration: /etc/rundeck/framework.properties
INFO  BootStrap: RSS feeds disabled
INFO  BootStrap: Preauthentication is disabled
INFO  BootStrap: Rundeck is ACTIVE: executions can be run.
INFO  BootStrap: Rundeck startup finished in 569ms
2017-03-06 14:52:01.350:INFO:/:main: Initializing Spring FrameworkServlet 'grails'
2017-03-06 14:52:01.383:INFO:oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@6d074b14{/,file:/var/lib/rundeck/exp/webapp/,AVAILABLE}{/var/lib/rundeck/exp/webapp}
2017-03-06 14:52:01.402:INFO:oejs.ServerConnector:main: Started ServerConnector@4a7b941f{HTTP/1.1}{0.0.0.0:4440}

```

**How to reproduce Behavior**

* Clear database
* Reinstall Rundeck (done via puppet in our environment)
* start rundeckd
* connect to the webinterface

**Comment by gkubok** [comment]
Hey @sumpfgottheit - I'm seeing the same issue running Rundeck 2.7.3-1 and PostgreSQL 9.6. Have you discovered what the issue is, or any other detrimental effects?

**Comment by bviktor** [comment]
I'm facing the same issue and the mentioned workaround doesn't work for me. Tried with the built-in driver,  `postgresql-42.0.0.jar` and `postgresql-9.4.1212.jre6.jar`, same result.

CentOS 7.3, PostgreSQL 9.6.2 from the official CentOS repo and Rundeck 2.8.0.1.29.GA.

Relevant config:

~~~
dataSource.driverClassName = org.postgresql.Driver
dataSource.url = jdbc:postgresql://localhost:5432/rundeck
dataSource.username=rundeckuser
#dataSource.password=nothing
~~~

Tried with or without :5432 in the url.

Relevant parts of `pg_hba.conf`:

~~~
host    all             all             127.0.0.1/32            trust
host    all             all             ::1/128                 trust
~~~

**Comment by sumpfgottheit** [comment]
Thank you for your findings. I won't be able to test it for the next 9 weeks...

**Comment by genadipost** [comment]
I have tried @gkubok workaround soi have replaced the JDBC driver with latest```postgresql-9.4.1212.jre6.jar ``` driver for Rundeck 2.7.3-1 and PostgreSQL 9.6, it worked great.

But after i upgraded to rundeck-2.8.0-1.29 (via yum) i'm getting the following error:

```
WARNING: HTTPS is not enabled, specify -Drundeck.ssl.config=/var/lib/rundeck/server/config/ssl.properties to enable.
2017-04-11 13:32:49.708:INFO:oejs.Server:main: jetty-9.0.7.v20131107
2017-04-11 13:32:52.211:INFO:oejw.StandardDescriptorProcessor:main: NO JSP Support for /, did not find org.apache.jasper.servlet.JspServlet
2017-04-11 13:32:53.467:INFO:/:main: Initializing Spring root WebApplicationContext
2017-04-11 13:33:08,445 [main] ERROR hbm2ddl.SchemaUpdate  - HHH000388: Unsuccessful: create table job_file_record (id int8 not null, version int8 not null, date_created timestamp not null, execution_id int8, expiration_date timestamp, file_name varchar(1024), file_state varchar(255) not null, job_id varchar(255) not null, last_updated timestamp not null, project varchar(255) not null, record_name varchar(255), record_type varchar(255) not null, server_nodeuuid text, sha varchar(64) not null, size int8 not null, storage_meta text, storage_reference text not null, storage_type varchar(255) not null, user varchar(255) not null, uuid varchar(255) not null, primary key (id))
2017-04-11 13:33:08,446 [main] ERROR hbm2ddl.SchemaUpdate  - ERROR: syntax error at or near "user"
  Position: 521
2017-04-11 13:33:08,482 [main] ERROR hbm2ddl.SchemaUpdate  - HHH000388: Unsuccessful: alter table job_file_record add constraint FK_dgig9thteq37sidkc38msxep foreign key (execution_id) references execution
2017-04-11 13:33:08,482 [main] ERROR hbm2ddl.SchemaUpdate  - ERROR: relation "job_file_record" does not exist
INFO  BootStrap: Starting Rundeck 2.8.0-1 (2017-04-11) ...
INFO  BootStrap: using rdeck.base config property: /var/lib/rundeck
INFO  BootStrap: loaded configuration: /etc/rundeck/framework.properties
INFO  BootStrap: RSS feeds disabled
INFO  BootStrap: Preauthentication is enabled
INFO  BootStrap: Rundeck is ACTIVE: executions can be run.
INFO  BootStrap: Rundeck startup finished in 677ms
2017-04-11 13:33:20.227:INFO:/:main: Initializing Spring FrameworkServlet 'grails'
2017-04-11 13:33:20.261:INFO:oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@5762658b{/,file:/var/lib/rundeck/exp/webapp/,AVAILABLE}{/var/lib/rundeck/exp/webapp}
2017-04-11 13:33:20.302:INFO:oejs.ServerConnector:main: Started ServerConnector@313a3af8{HTTP/1.1}{0.0.0.0:4440}
ERROR SqlExceptionHelper: ERROR: relation "job_file_record" does not exist
  Position: 695
```

**Comment by andsantrj** [comment]
@genadipost The error is in your log... the issue is, the rundeck is trying to create a column called user, but this is a reserved word ... you need to execute the create table enclosing with double quotes "user".

It will work fine.

**Comment by gschueler** [comment]
Rundeck 2.8 added a new table with a `user` column, that can be a problem if it is a reserved word, we will have to map to another name

**Comment by PawelRybak** [comment]
I think there are two separate issues discussed here:
one is using "user" (unquoted) as a column name - this is fixed from what I can see (RD 2.8.2, PG 9.6.2):

```
\d job_file_record
               Table "public.job_file_record"
      Column       |            Type             | Modifiers 
-------------------+-----------------------------+-----------
 id                | bigint                      | not null
 version           | bigint                      | not null
 date_created      | timestamp without time zone | not null
 execution_id      | bigint                      | 
 expiration_date   | timestamp without time zone | 
 file_name         | character varying(1024)     | 
 file_state        | character varying(255)      | not null
 job_id            | character varying(255)      | not null
 last_updated      | timestamp without time zone | not null
 project           | character varying(255)      | not null
 record_name       | character varying(255)      | 
 record_type       | character varying(255)      | not null
 server_nodeuuid   | text                        | 
 sha               | character varying(64)       | not null
 SIZE              | bigint                      | not null
 storage_meta      | text                        | 
 storage_reference | text                        | not null
 storage_type      | character varying(255)      | not null
 rduser            | character varying(255)      | not null
 uuid              | character varying(255)      | not null
```

the column name is "rduser" now.

However the originally reported issue:

```
ERROR: column am.amcanorder does not exist Position: 407
```

is still present.
in version 9.6 Postgres modified the `pg_catalog.pg_am` [table](https://www.postgresql.org/docs/9.6/static/catalog-pg-am.html) and column `amcanorder` does not exist anymore.

The error is thrown when executing this query:

```sql
SELECT
    NULL AS TABLE_CAT,
    n.nspname AS TABLE_SCHEM,
    ct.relname AS TABLE_NAME,
    NOT i.indisunique AS NON_UNIQUE,
    NULL AS INDEX_QUALIFIER,
    ci.relname AS INDEX_NAME,
    CASE
        i.indisclustered 
        WHEN
            true 
        THEN
            1 
        ELSE
            CASE
                am.amname 
                WHEN
                    'hash' 
                THEN
                    2 
                ELSE
                    3 
            END
    END
    AS TYPE, 
    (
        i.keys
    )
    .n AS ORDINAL_POSITION, pg_catalog.pg_get_indexdef(ci.oid, 
    (
        i.keys
    )
    .n, false) AS COLUMN_NAME, 
    CASE
        am.amcanorder 
        WHEN
            true 
        THEN
            CASE
                i.indoption[(i.keys).n - 1] & 1 
                WHEN
                    1 
                THEN
                    'D' 
                ELSE
                    'A'
            END
            ELSE
                NULL 
    END
    AS ASC_OR_DESC, ci.reltuples AS CARDINALITY, ci.relpages AS PAGES, pg_catalog.pg_get_expr(i.indpred, i.indrelid) AS FILTER_CONDITION 
FROM
    pg_catalog.pg_class ct 
    JOIN
        pg_catalog.pg_namespace n 
        ON (ct.relnamespace = n.oid) 
    JOIN
        (
            SELECT
                i.indexrelid,
                i.indrelid,
                i.indoption,
                i.indisunique,
                i.indisclustered,
                i.indpred,
                i.indexprs,
                information_schema._pg_expandarray(i.indkey) AS keys 
            FROM
                pg_catalog.pg_index i
        )
        i 
        ON (ct.oid = i.indrelid) 
    JOIN
        pg_catalog.pg_class ci 
        ON (ci.oid = i.indexrelid) 
    JOIN
        pg_catalog.pg_am am 
        ON (ci.relam = am.oid) 
WHERE
    true 
    AND n.nspname = 'public' 
    AND ct.relname = 'auth_token' 
ORDER BY
    NON_UNIQUE,
    TYPE,
    INDEX_NAME,
    ORDINAL_POSITION;
```

I suspect that with Postgres 9.6 the query should use `pg_indexam_has_property(ci.relam, 'can_order')` instead of `am.amcanorder`.

**Comment by gschueler** [comment]
that must be a problem with the postgres jdbc driver, or with Hibernate, I think

**Comment by PawelRybak** [comment]
no, table `pg_catalog.pg_am` has changed:

in PG 9.5

```
\d pg_catalog.pg_am
        Table "pg_catalog.pg_am"
     Column      |   Type   | Modifiers 
-----------------+----------+-----------
 amname          | name     | not null
 amstrategies    | smallint | not null
 amsupport       | smallint | not null
 amcanorder      | boolean  | not null
 amcanorderbyop  | boolean  | not null
 amcanbackward   | boolean  | not null
 amcanunique     | boolean  | not null
 amcanmulticol   | boolean  | not null
 amoptionalkey   | boolean  | not null
 amsearcharray   | boolean  | not null
 amsearchnulls   | boolean  | not null
 amstorage       | boolean  | not null
 amclusterable   | boolean  | not null
 ampredlocks     | boolean  | not null
 amkeytype       | oid      | not null
 aminsert        | regproc  | not null
 ambeginscan     | regproc  | not null
 amgettuple      | regproc  | not null
 amgetbitmap     | regproc  | not null
 amrescan        | regproc  | not null
 amendscan       | regproc  | not null
 ammarkpos       | regproc  | not null
 amrestrpos      | regproc  | not null
 ambuild         | regproc  | not null
 ambuildempty    | regproc  | not null
 ambulkdelete    | regproc  | not null
 amvacuumcleanup | regproc  | not null
 amcanreturn     | regproc  | not null
 amcostestimate  | regproc  | not null
 amoptions       | regproc  | not null
```

in PG 9.6

```
\d pg_catalog.pg_am
    Table "pg_catalog.pg_am"
  Column   |  Type   | Modifiers 
-----------+---------+-----------
 amname    | name    | not null
 amhandler | regproc | not null
 amtype    | "char"  | not null
```

hope this helps

**Comment by gschueler** [comment]
Rundeck does not query that table, so what i mean is that whatever is creating the (broken) query must not be compatible with PG 9.6.  That could be jdbc driver being used.  Is there a newer one available?

**Comment by PawelRybak** [comment]
Ouch, I did not get that, sorry.
I dug this query up from Postgres' logs so I did not think about the libraries/drivers.
This it the error spilled by Rundeck if it may be of any help:

```
2017-04-24 14:41:13,615 [main] ERROR spi.SqlExceptionHelper  - ERROR: column am.amcanorder does not exist
  Position: 407
2017-04-24 14:41:13,616 [main] ERROR hbm2ddl.SchemaUpdate  - HHH000299: Could not complete schema update
java.lang.NullPointerException
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:126)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:112)
	at org.hibernate.tool.hbm2ddl.DatabaseMetadata.getTableMetadata(DatabaseMetadata.java:158)
	at org.hibernate.cfg.Configuration.generateSchemaUpdateScriptList(Configuration.java:1202)
	at org.hibernate.tool.hbm2ddl.SchemaUpdate.execute(SchemaUpdate.java:212)
	at org.hibernate.tool.hbm2ddl.SchemaUpdate.execute(SchemaUpdate.java:178)
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:522)
	at org.hibernate.cfg.Configuration.buildSessionFactory(Configuration.java:1857)
	at org.codehaus.groovy.grails.orm.hibernate.cfg.GrailsAnnotationConfiguration.buildSessionFactory(GrailsAnnotationConfiguration.java:372)
	at org.codehaus.groovy.grails.orm.hibernate.ConfigurableLocalSessionFactoryBean.doBuildSessionFactory(ConfigurableLocalSessionFactoryBean.java:479)
	at org.codehaus.groovy.grails.orm.hibernate.ConfigurableLocalSessionFactoryBean.buildSessionFactory(ConfigurableLocalSessionFactoryBean.java:473)
	at org.codehaus.groovy.grails.orm.hibernate.ConfigurableLocalSessionFactoryBean.afterPropertiesSet(ConfigurableLocalSessionFactoryBean.java:387)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1613)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1550)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:539)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:475)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:228)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:298)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:193)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:328)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveValueIfNecessary(BeanDefinitionValueResolver.java:108)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyPropertyValues(AbstractAutowireCapableBeanFactory.java:1457)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1198)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:537)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:475)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:228)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:298)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:198)
	at org.codehaus.groovy.grails.orm.support.TransactionManagerPostProcessor.initialize(TransactionManagerPostProcessor.java:74)
	at org.codehaus.groovy.grails.orm.support.TransactionManagerPostProcessor.setBeanFactory(TransactionManagerPostProcessor.java:52)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeAwareMethods(AbstractAutowireCapableBeanFactory.java:1573)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1541)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:539)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:475)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:228)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:298)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:198)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.registerBeanPostProcessors(PostProcessorRegistrationDelegate.java:199)
	at org.springframework.context.support.AbstractApplicationContext.registerBeanPostProcessors(AbstractApplicationContext.java:620)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:467)
	at org.codehaus.groovy.grails.commons.spring.DefaultRuntimeSpringConfiguration.getApplicationContext(DefaultRuntimeSpringConfiguration.java:156)
	at org.codehaus.groovy.grails.commons.spring.GrailsRuntimeConfigurator.initializeContext(GrailsRuntimeConfigurator.java:188)
	at org.codehaus.groovy.grails.commons.spring.GrailsRuntimeConfigurator.configure(GrailsRuntimeConfigurator.java:168)
	at org.codehaus.groovy.grails.commons.spring.GrailsRuntimeConfigurator.configure(GrailsRuntimeConfigurator.java:127)
	at org.codehaus.groovy.grails.web.context.GrailsConfigUtils.configureWebApplicationContext(GrailsConfigUtils.java:126)
	at org.codehaus.groovy.grails.web.context.GrailsContextLoaderListener.initWebApplicationContext(GrailsContextLoaderListener.java:109)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:106)
	at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:788)
	at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:434)
	at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:780)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:284)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1322)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:732)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:490)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:69)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:118)
	at org.eclipse.jetty.server.Server.start(Server.java:342)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:100)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:60)
	at org.eclipse.jetty.server.Server.doStart(Server.java:290)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:69)
	at com.dtolabs.rundeck.RunServer.run(RunServer.java:119)
	at com.dtolabs.rundeck.RunServer.main(RunServer.java:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.dtolabs.rundeck.ExpandRunServer.invokeMain(ExpandRunServer.java:752)
	at com.dtolabs.rundeck.ExpandRunServer.execute(ExpandRunServer.java:709)
	at com.dtolabs.rundeck.ExpandRunServer.run(ExpandRunServer.java:309)
	at com.dtolabs.rundeck.ExpandRunServer.main(ExpandRunServer.java:113)
```

**Comment by PawelRybak** [comment]
@gschueler - yes, you are perfectly right, this query comes from Postgres' jdbc.
Thanks and sorry for trouble.

**Comment by gschueler** [comment]
@PawelRybak no problem. For others: is there an upgrade to the jdbc driver for 9.6?

**Comment by PawelRybak** [comment]
no, 42.0.0 is the latest at the moment;
I have opened an issue in pgjdbc [project](https://github.com/pgjdbc/pgjdbc/issues/815)

it would be nice though if someone could try to reproduce that - maybe it is just my screw up :(

**Comment by PawelRybak** [comment]
jFYI: I just created confusion, nothing more.

I copied the latest Postgres JDBC into `$RDECK_BASE/server/lib/` (don't ask why ... ;) ) and Rundeck used the old driver shipped with the package (`9.1` version), not (as I thought) `42.0.0`

After moving it to `$RDECK_BASE/server/exp/webapp/WEB-INF/lib/` the error (`column am.amcanorder does not exist`) disappeared .

Sorry about that.
